src/OrbitGl/App.cpp

```diff
@@ -999,121 +1002,130 @@ void OrbitApp::SendErrorToUi(const std::string& title, const std::string& text)
-                                    module_data->file_path(), debug_file_path));
-                SCOPED_TIMED_LOG("Copying \"%s\"", debug_file_path);
-                auto scp_result =
-                    secure_copy_callback_(debug_file_path, local_debug_file_path.string());
-                if (!scp_result) {
-                  SendErrorToUi(
-                      "Error loading symbols",
-                      absl::StrFormat("Could not copy debug info file from the remote: %s",
-                                      scp_result.error().message()));
-                  modules_currently_loading_.erase(module_data->file_path());
-                  return;
-                }
-              }
+        const std::filesystem::path local_debug_file_path =
+            symbol_helper_.GenerateCachedFileName(module_file_path);
+
+        scoped_status.UpdateMessage(
+            absl::StrFormat(R"(Copying debug info file for "%s" from remote: "%s"...)",
+                            module_file_path, debug_file_path));
+        SCOPED_TIMED_LOG("Copying \"%s\"", debug_file_path);
+        auto scp_result = secure_copy_callback_(debug_file_path, local_debug_file_path.string());
+        if (!scp_result) {
+          return ErrorMessage{absl::StrFormat("Could not copy debug info file from the remote: %s",
+                                              scp_result.error().message())};
+        }
 
-              LoadSymbols(local_debug_file_path, module_data, std::move(function_hashes_to_hook),
-                          std::move(frame_track_function_hashes));
-            });
+        return local_debug_file_path;
       });
 }
 
-void OrbitApp::LoadModules(
-    const std::vector<ModuleData*>& modules,
-    absl::flat_hash_map<std::string, std::vector<uint64_t>> function_hashes_to_hook_map,
-    absl::flat_hash_map<std::string, std::vector<uint64_t>> frame_track_function_hashes_map) {
+orbit_base::Future<void> OrbitApp::LoadModules(absl::Span<const ModuleData* const> modules) {
+  std::vector<orbit_base::Future<void>> futures;
+  futures.reserve(modules.size());
+
   for (const auto& module : modules) {
-    if (modules_currently_loading_.contains(module->file_path())) {
-      continue;
-    }
-    modules_currently_loading_.insert(module->file_path());
+    futures.emplace_back(LoadModule(module));
+  }
 
-    std::vector<uint64_t> function_hashes_to_hook;
-    if (function_hashes_to_hook_map.contains(module->file_path())) {
-      function_hashes_to_hook = function_hashes_to_hook_map.at(module->file_path());
-    }
+  return orbit_base::JoinFutures(futures);
+}
 
-    std::vector<uint64_t> frame_track_function_hashes;
-    if (frame_track_function_hashes_map.contains(module->file_path())) {
-      frame_track_function_hashes = frame_track_function_hashes_map.at(module->file_path());
-    }
+orbit_base::Future<void> OrbitApp::LoadModule(const ModuleData* module) {
```

Comment:
I actually dont really know how to make this better. Am I assuming correctly, that you did the local variable lambdas to avoid deep nesting with `.then`?